README.rst
******************************
processr - process your dicts!
******************************


.. image:: https://img.shields.io/travis/entropiae/processr.svg
    :target: https://travis-ci.org/entropiae/processr
    :alt: Travis-CI

WIP. If you're interested, reach me on `twitter <https://twitter.com/entropiae>`_.


HISTORY.rst
=======
History
=======

0.1.0 (2016-4-6)
------------------

* First release on PyPI.


tests/test_stages.py
# -*- coding: utf-8 -*-

from processr.processr import (
    rename_keys,
    project_dict,
    transform_values,
    transform_dict,
    transform_values_strict)

import pytest


##############################################################
#                        rename_keys                         #
##############################################################

def test_rename_keys():
    d = {'the_answer': 42}
    opts = {'the_answer': 'the_universe_and_everything'}
    expected_output = {'the_universe_and_everything': 42}

    output = rename_keys(d, opts)
    assert output == expected_output


def test_rename_keys_extra_fields():
    d = {'the_answer': 42, 'not_the_answer': 43}
    opts = {'the_answer': 'the_universe_and_everything'}
    expected_output = {'the_universe_and_everything': 42, 'not_the_answer': 43}

    output = rename_keys(d, opts)
    assert output == expected_output


##############################################################
#                          project                           #
##############################################################

def test_project_dict():
    d = {'the_answer': 42, 'not_the_answer': 43}
    opts = ('the_answer', )
    expected_output = {'the_answer': 42}

    output = project_dict(d, opts)
    assert output == expected_output


def test_project_dict_extra_fields():
    d = {'the_answer': 42, 'not_the_answer': 43}
    opts = ('the_answer', 'the_universe')

    with pytest.raises(KeyError):
        project_dict(d, opts)


##############################################################
#                      transform_values                      #
##############################################################

def test_transform_values():
    d = {'not_the_answer': [9, 9, 9, 9, 9, 9]}
    opts = {'not_the_answer': [sum, str]}

    # I always thought something was fundamentally wrong with the universe
    expected_output = {'not_the_answer': '54'}
    output = transform_values(d, opts)
    assert output == expected_output


def test_transform_values_extra_fields():
    d = {'not_the_answer': [9, 9, 9, 9, 9, 9], 'the_answer': 42}
    opts = {'not_the_answer': [sum, str]}

    expected_output = {'not_the_answer': '54', 'the_answer': 42}
    output = transform_values(d, opts)
    assert output == expected_output


def test_transform_values_non_strictness():
    d = {'not_the_answer': [9, 9, 9, 9, 9, 9]}
    opts = {'the_answer': [str]}

    expected_output = {'not_the_answer': [9, 9, 9, 9, 9, 9]}
    output = transform_values(d, opts)
    assert output == expected_output


def test_transform_values_strict():
    d = {'not_the_answer': [9, 9, 9, 9, 9, 9]}
    opts = {'not_the_answer': [sum, str]}

    expected_output = {'not_the_answer': '54'}
    output = transform_values_strict(d, opts)
    assert output == expected_output


def test_transform_values_extra_fields():
    d = {'not_the_answer': [9, 9, 9, 9, 9, 9], 'the_answer': 42}
    opts = {'not_the_answer': [sum, str]}

    expected_output = {'not_the_answer': '54', 'the_answer': 42}
    output = transform_values_strict(d, opts)
    assert output == expected_output


def test_transform_values_strict_ko():
    d = {'not_the_answer': [9, 9, 9, 9, 9, 9]}
    opts = {'the_answer': [str]}

    with pytest.raises(KeyError):
        transform_values_strict(d, opts)

##############################################################
#                       transform_dict                       #
##############################################################

def test_dict_transform():
    d = {'the_answer': 42, 'not_the_answer': 43}
    ops = [
        lambda d: dict((v, k) for k, v in d.items())  # Invert key and values
    ]

    expected_output = {42: 'the_answer', 43: 'not_the_answer'}
    output = transform_dict(d, ops)
    assert output == expected_output


tests/test_processr.py
# -*- coding: utf-8 -*-

from processr.processr import process_value


def test_process_simple():
    input_value = 42
    opts = str
    expected_output = '42'

    output = process_value(input_value, opts)
    assert output == expected_output


def test_process_value_list():
    input_value = [1 for _ in range(42)]
    opts = [sum, str]
    expected_output = '42'

    output = process_value(input_value, opts)
    assert output == expected_output


def test_process_value_fun_with_args():
    input_value = 41

    def sum(a, b):
        return a + b

    opts = [(sum, [], {'b': 1}), str]
    expected_output = '42'

    output = process_value(input_value, opts)
    assert output == expected_output


tests/test_transformers.py
# -*- coding: utf-8 -*-

import pytest

from processr.transformers import set_value


def test_set_value():
    provided_input = {}
    expected_output = {'the_answer': 42}

    output = set_value(
        provided_input,
        'the_answer',
        42
    )
    assert output == expected_output


def test_set_value_callable():
    provided_input = {}
    expected_output = {'the_answer': 42}

    output = set_value(
        provided_input,
        'the_answer',
        lambda _: 42,
    )
    assert output == expected_output


def test_set_value_if():
    provided_input = {}
    expected_output = {'the_answer': 42}

    output = set_value(
        provided_input,
        'the_answer',
        42,
        lambda value: value % 2 == 0
    )
    assert output == expected_output


def test_set_value_if_not():
    provided_input = {}
    expected_output = {}

    output = set_value(
        provided_input,
        'the_answer',
        42,
        lambda value: value == 6 * 9
    )
    assert output == expected_output


def test_set_value_if_exception_bubbling():
    provided_input = {}

    def raise_exc(_):
        raise TypeError

    with pytest.raises(TypeError):
        set_value(
            provided_input,
            'the_answer',
            raise_exc
        )


processr/transformers.py
# -*- coding: utf-8 -*-

from __future__ import absolute_import

from processr.compat import reduce

from processr.compat import abc


##############################################################
#                 Field transformers & utils                 #
##############################################################

def apply_default(value, default):
    return value if value is not None else default


def apply_map(fun):
    """
    Restituisce una funzione che trasforma gli elementi di una sequenza in modo
    lazy usando la fun fornita
    """
    def _map(sequence):
        return (fun(item) for item in sequence)
    return _map


def apply_filter(fun):
    """
    Restituisce una funzione che filtra in modo lazy gli elementi di una
    sequenza
    """
    def _filter(sequence):
        return (item for item in sequence if fun(item))
    return _filter


##############################################################
#                     Dict transformers                      #
##############################################################

def set_value(d, key, value, func=lambda _: True):
    """
    Add (or replace) a field to a dict only if the result of `func` is True
    for the computed value.
    Value could be a plain value or a callable. In the latter case,
    it will be called (with row as an argument) to build the real value.

    Add it to your pipeline like:
    (
        set_value
        {
            'key': 'the_answer',
            'value': 42,  # or something like lambda d: 42
            'func': lambda x: True  # optional
        }
    )

    :param d: the input dictionary
    :param key: the key of the value to set
    :param value: the value to set; could be a callable
    :param func: an optional boolean function that will be called
    on the value to check if should be added to the dictionary
    :return: a dictionary
    """

    if isinstance(value, abc.Callable):
        # Q: Why are you using LBYL instead of EAFP?
        # A: Catching the TypeError raised by value() when its not
        # a callable will swallow potential exceptions raised
        # by the code inside it when it is a callable.
        value = value(d)

    if func(value):
        d[key] = value

    return d


def get_value(row, source_key, default=None):
    try:
        value = reduce(lambda value, key: value[key], source_key.split('.'), row)
    except KeyError:
        value = default
    return value


def flatten(row, source_key, destination_keys):
    """
    Convert a list of dicts (identified by the source_key key in the input row) in a dict containing
    (destination_key, list) items.

    # Example
    row = {
        'doc_key': [
            {
                'sub_key_1': 1,
                'sub_key_2': 2
            },
            {
                'sub_key_1': 3
                'sub_key_2': 4
            }
        ]
    }

    row = flatten(row, 'doc_key', ['sub_key_1', 'sub_key_2'])

    output = {
        'sub_key_1': [1, 3],
        'sub_key_2': [2, 4],
    }
    """
    subdocuments = row.pop(source_key, [])

    def zip_subdocuments(key):
        zipped_value = (subdocument.get(key, None) for subdocument in subdocuments)
        return [value for value in zipped_value if value is not None]

    flatten_fields = [(key, zip_subdocuments(key)) for key in destination_keys]
    return dict(row.items() + flatten_fields)


def copy_value(row, source_key, destination_key, default=None):
    return set_value(row, destination_key, get_value(row, source_key, default))











def passthrough_on_exception(*exceptions):
    """
    Wrap a row transformer, catching exceptions throwed by it.
    If the exception is among the ones specified, it return the input row value.
    """
    def wrapper(f):
        def wrapped(arg):
            try:
                return f(arg)
            except exceptions:
                return arg
        return wrapped
    return wrapper


processr/processr.py
# -*- coding: utf-8 -*-

from __future__ import absolute_import

import logging
from itertools import chain

from processr.compat import reduce, abc, NullHandler

# initialize log & set default logging handler
# to avoid 'No handler found' warnings.
log = logging.getLogger(__name__)
log.addHandler(NullHandler())


##############################################################
#                           Stages                           #
##############################################################


def rename_keys(d, stage_opts):
    """
    Returns a dictionary composed by items from `d`; when a key is found
    in `opts`, the corresponding value is used as key in the new dictionary.

    >>> rename_keys({'key': 42}, {'key': 'new_key'})
    {'new_key': 42}

    :param d: the input dictionary
    :param stage_opts: a mapping used to translate old_key -> new_key
    :return: a dictionary
    """
    return dict((stage_opts.get(k, k), v) for k, v in d.items())


def project_dict(d, stage_opts):
    """
    Returns a dictionary composed by items from `d` whose keys are
    in `opts` collection.
    If `opts`contains a key which is not in `d` a `KeyError` is raised.

    >>> project_dict({'a': 1, 'b': 2}, ('a', ))
    {'a': 1}

    will be maintained in the returned dict.
    :param d: the input dictionary
    :param stage_opts: a collection containing the key which
    :return: a dictionary
    """
    return dict((k, d[k]) for k in stage_opts)


def transform_values(d, stage_opts):
    """
    Return a new dictionary whose values are taken from `d` and processed
    using the list of callable having the same key in `stage_opts`.
    Values with no specified transformers are copied untouched.

    >>> transform_values({'the_answer': [41, 1]}, {'the_answer': [sum, str]})
    {'the_answer': '42'}

    :param d: the input dictionary
    :param stage_opts: a dictionary containing key -> list of callables
    :return: a dictionary
    """
    return dict(
        (key, process_value(value, stage_opts.get(key, [])))
        for key, value in d.items()
    )


def transform_values_strict(d, stage_opts):
    """
    Like `transform_values`, but raise a `KeyError` if a key from stage_opts
    isn't found in d.
    :param d: the input dictionary
    :param stage_opts: a dictionary containing key -> list of callables
    :return: a dictionary
    """

    unchanged_items = (
        (key, value)
        for key, value in d.items()
        if key not in stage_opts
    )

    processed_items = (
        (key, process_value(d[key], opts))
        for key, opts in stage_opts.items()
    )

    return dict(
        chain(unchanged_items, processed_items)
    )


def transform_dict(d, stage_opts):
    """
    Return a new dictionary built applying all callable in `opts` to `d`.

    >>> reverse_dict = lambda d: dict((v, k) for k, v in d.items())
    >>> transform_dict({'the_answer': 42}, [reverse_dict])
    {42: 'the_answer'}

    :param d: the input dictionary
    :param stage_opts: a list of callable to apply to the input dictionary
    :return: a dictionary
    """
    return process_value(d, stage_opts)


class StageDefinitions(dict):
    """
    Provides a way to get a stage handler by its name, and a way
    to trivially add custom stages.

    :param add_default_stages: add the 4 default stage handlers
    """

    _default_stages = {
        'project_dict': project_dict,
        'rename_keys': rename_keys,
        'transform_values': transform_values,
        'transform_dict': transform_dict,
        'transform_values_strict': transform_values_strict
    }

    def __init__(self, add_default_stages=True):

        super(StageDefinitions, self).__init__()

        if add_default_stages:
            for stage_name, stage_handler in self._default_stages.items():
                self[stage_name] = [stage_handler]


##############################################################
#                     Process all the things!                #
##############################################################

class InvalidTransformerFormat(Exception):
    pass


def process_value(value, fs):
    """
    Process a value.

    :param value: the input value to process
    :param fs: transformer(s) used to process value.
        Could be a callable, a (callable, args, kwargs) tuple
        or a list of both.
    :return: the processed value
    """
    if isinstance(fs, tuple):
        # The transformer is a function which requires extra arguments,
        # so is expressed as a (f, args, kwargs) tuple.
        f, kwargs = fs
        log.debug(
            {'transformer': f, 'kwargs': kwargs, 'input': value}
        )
        return_value = f(value, **kwargs)
        log.debug({'output': return_value})
    elif isinstance(fs, abc.Iterable):
        # The transformers is actually a list of transformers.
        # Recursively call process_value to apply all of them.
        return_value = reduce(process_value, fs, value)
    elif isinstance(fs, abc.Callable):
        # Transformer is a callable w/o extra arguments.
        # Call it!
        log.debug({'transformer': fs, 'input': value})
        return_value = fs(value)
        log.debug({'output': return_value})
    else:
        raise InvalidTransformerFormat(fs)
    return return_value


def process(d, pipeline, stage_definitions=StageDefinitions()):
    """
    Process a dictionary according to the given pipeline, using
    the stage handlers defined in stage_definitions.

    :param d: the dictionary to process
    :param pipeline: the processing pipeline
    :param stage_definitions: a (stage_name, stage_handler) mapping
    :return: a dictionary
    """
    def process_stage(d, stage):
        stage_name, stage_option = stage
        try:
            handler = stage_definitions[stage_name]
        except KeyError:
            raise
        return handler(d, stage_option)

    return reduce(process_stage, pipeline, d)
